### DISCUSSION TITLE
Help with Registering a Foliage Placer Type

### DISCUSSION THREAD

**Discussion by ReikanFurrow** [FabricMC/fabric-api#2428] [Mod Dev Support] [answered]
I've been trying to make my own custom foliage shapes for a palm tree, but I can't seem to figure out a way to register my own foliageplacertypes. Any insight would be appreciated.
![image](https://user-images.githubusercontent.com/34829379/181429793-02157616-5ade-4ab6-bd83-1118ddd37722.png)
I apologize in advance if the answer is obvious, I am somewhat new to Java and am using this mod as a vehicle to learn.

Edit: some information that might be important: I am trying to create a palm tree with my own custom foliage shape for it. In the  Tree Configured Feature, it calls for a FoliagePlacer, and I'd like to create my own.

**Comment by Daomephsta** [comment, marked as answer]
The constructor of `FoliagerPlacerType` is private, making it inaccessible to any code outside of `FoliagerPlacerType.java`. 
There are 2 ways to get around this. Both are Fabric or Minecraft specific, so it's unsurprising you couldn't find them, search engines don't give great results for Minecraft modding.

Option 1 is creating a public Mixin invoker for the `FoliagerPlacerType` constructor.
<https://fabricmc.net/wiki/tutorial:mixin_accessors>
<https://jenkins.liteloader.com/view/Other/job/Mixin/javadoc/org/spongepowered/asm/mixin/gen/Invoker.html>  

Option 2 is using Fabric's access wideners to change the access modifier of the constructor.
<https://www.fabricmc.net/wiki/tutorial:accesswideners>

> I apologize in advance if the answer is obvious, I am somewhat new to Java and am using this mod as a vehicle to learn.

This is fine as long as you have prior programming experience. Otherwise it's better to learn Java first, Minecraft modding is a difficult learning environment.

**Reply by ReikanFurrow** [reply]
Thank you for the swift and well composed response. I will use option 2, as it seems to be the solution to a number of other problems I have been facing.
